registry/storage/tagstore.go

```diff
@@ -134,44 +152,118 @@ func (ts *tagStore) linkedBlobStore(ctx context.Context, tag string) *linkedBlob
 
 // Lookup recovers a list of tags which refer to this digest.  When a manifest is deleted by
 // digest, tag entries which point to it need to be recovered to avoid dangling tags.
-func (ts *tagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
+func (ts *TagStore) Lookup(ctx context.Context, desc distribution.Descriptor) ([]string, error) {
 	allTags, err := ts.All(ctx)
 	switch err.(type) {
+	case nil:
+		break
 	case distribution.ErrRepositoryUnknown:
 		// This tag store has been initialized but not yet populated
 		break
-	case nil:
-		break
 	default:
 		return nil, err
 	}
 
-	var tags []string
-	for _, tag := range allTags {
-		tagLinkPathSpec := manifestTagCurrentPathSpec{
-			name: ts.repository.Named().Name(),
-			tag:  tag,
+	lookupErr := &atomicError{}
+
+	allTagsCount := len(allTags)
+	inputChan := make(chan string)
+	outputChan := make(chan string, allTagsCount)
+
+	waitGroup := sync.WaitGroup{}
+	waitGroup.Add(allTagsCount)
```

Comment:
There is no need to add `allTagsCount` to the `WaitGroup` here. We should just call `waitGroup.Add(1)` every time we launch a new goroutine (down below).